src/engine/SCons/Tool/link.py

```diff
@@ -81,6 +81,8 @@ def _lib_emitter(target, source, env, **kw):
     if Verbose:
         print("_lib_emitter: target[0]={!r}".format(target[0].get_path()))
     for tgt in target:
+        if SCons.Util.is_String(tgt):
```

Comment:
Curious why this started failing now?

Reply:
I don't know either, I don't know the last time we had the test/Win32/mingw.py test passing to see what changed since then.

Reply:
A bit of a weakness of our tests which just skip a test if the tool/setup is not there is if no CI environment is set up for a particular config, it just doesn't get tested. We have potential exposure on mingw, D, docbook, TeX, SWIG, packaging.  Some of these have gotten added in recently, others fail to be added (e.g. the WiX toolset for msi packaging: apparently that code has rotted, or WiX has changed out from underneath it).

Reply:
Indeed.  Some tests will create fake tools to run with always and/or if the tool doesn't exist in the environment.  However in some cases that's insufficient to test if the scons code works.  Ideally we'd get a list of all tools missing and/or causing tests to be skipped at the end of a test run.. but that could be considerable work to make happen.

Any thoughts on how to better address?

Reply:
Yup. dev list.. :)

Reply:
Curious though this issue should've affected all linkers and not just mingw... if they all needed to be turned into nodes at this point?  Does mingw fail if you don't do this bit?

Reply:
Yeah, it will get an exception that 'str' has no member 'attributes' just below. I'm not sure if there somewhere further up the chain that the target should have been converted.

Reply:
Still should be some kind of check here even if the real fix is somewhere else. Maybe a try/except? Add a warning message? I think assuming its a file if its a string is a pretty good default case.

Reply:
Yes. Can you change to try/except and put a comment in the except that we'd expect the target to be a Node at this point.  And a # TODO to track down the why at some point?
I don't want to lose this discovery without some reasonable way of remembering to address it at some point if possible.

Reply:
I chased this down and found that mingw is getting an emitter list of [mslink.py winLibEmitter, link.py shlib_emitter, mingw.py shlib_emitter]. mslink's winLibEmitter never converts its new targets to File's and just leaves them as strings:
https://github.com/SCons/scons/blob/9f090132091e56c123262106f3563142787c2fcf/src/engine/SCons/Tool/mslink.py#L144-L154
The other tool with emitters take care to convert targets to File objects:
https://github.com/SCons/scons/blob/9f090132091e56c123262106f3563142787c2fcf/src/engine/SCons/Tool/cyglink.py#L100-L106
So the fix and be applied by converting the the targets in mslink.py's winLibEmitter to File's. Probably still want to have a check in link.py to convert str's as well.

Reply:
I left the if check instead of a try except because the current code looks cleaner than:
```python
 for tgt in target:
        try:
            tgt.attributes.shared = 1
        except AttributeError:
            tgt = env.File(tgt)
            tgt.attributes.shared = 1
```
I am not sure about the speed up. Let me know your opinion.